### PR TITLE
HPCC-30949 Ensure that duplicate statistics are not added to a CStatsCollection

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jlib/jstats.cpp [hunk 1]
```diff
@@ -1919,6 +1919,10 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
 //other public interface functions
     void addStatistic(StatisticKind kind, unsigned __int64 value)
     {
+#if 1 //def _DEBUG
```
Comment (ghalliday) [inline system/jlib/jstats.cpp:1922]: This will be debug only in the final PR - pushed to run the full regression suite.

Comment (shamser) [reply system/jlib/jstats.cpp:1922]: It would be useful to have a comment to say that addStatistic shouldn't ever add the same kind multiple times.

**Review by shamser — APPROVED** [review]
I think a comment on addStatistic would be useful.

**Comment by ghalliday** [comment]
@shamser please sanity check the comments and the change to the #if

File: system/jlib/jstats.cpp [hunk 2]
```diff
@@ -1917,8 +1917,15 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
     }
 
 //other public interface functions
+    // addStatstic should only be used if it is known that the kind is not already there
```
Comment (shamser) [inline system/jlib/jstats.cpp:1920]: minor: spelling addStatistic

File: roxie/ccd/ccdserver.cpp [hunk 3]
```diff
@@ -390,6 +390,60 @@ class IndirectAgentContext : implements IRoxieAgentContext, public CInterface
...
+    {
+        gatherer->beginChannelScope(id);
+    }
+    virtual void endScope() override
+    {
+        gatherer->endScope();
+    }
+    virtual void addStatistic(StatisticKind kind, unsigned __int64 value) override
+    {
+        //Always merge rather than updating
```
Comment (shamser) [inline roxie/ccd/ccdserver.cpp:430]: minor: merge rather than "add"

**Review by shamser — APPROVED** [review]
Couple of minors related to comments.

**Review by richardkchapman — APPROVED** [review]
Looks plausible

**Comment by ghalliday** [comment]
Fixed comments and force pushed.

**Pull request merged**